File: 이코테 복습(python, c++, java)/다이나믹 프로그래밍/java/효율적인_화폐_구성.java
package javastudy.이코테.다이나믹프로그래밍;

import java.util.Arrays;
import java.util.Scanner;

public class 효율적인_화폐_구성 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        int[] coins = new int[n];
        for (int i = 0; i < n; i++) {
            coins[i] = sc.nextInt();
        }

        int[] dp = new int[m + 1];
        Arrays.fill(dp, 10001);
        dp[0] = 0;

        for (int coin : coins) {
            for (int i = coin; i < m + 1; i++) {
                if (dp[i - coin] != 10001) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }

        if (dp[m] == 10001) {
            System.out.println(-1);
        } else {
            System.out.println(dp[m]);
        }
    }
}

File: 이코테 복습(python, c++, java)/다이나믹 프로그래밍/python/효율적인 화폐 구성.py
N, M = map(int, input().split())
coins = []
for _ in range(N):
    coins.append(int(input()))
dp = [1e9] * 10001
dp[0] = 0
for coin in coins:
    dp[coin] = 1
    for i in range(coin + 1, M + 1):
        if dp[i - coin] != 1e9:
            dp[i] = min(dp[i], dp[i - coin] + 1)
if dp[M] == 1e9:
    print(-1)
else:
    print(dp[M])

